### PR TITLE
Make PullSchema work and add ability to delete

### LINKED ISSUE #146: Reimplement PullSchema streaming
This was removed when we eliminated `agent.go`.

### PR DESCRIPTION
**Pull request by wesbillman**
Fixes #146 
https://github.com/TBD54566975/ftl/assets/51647/8cbee1d6-071c-4495-bd51-d7c092f50044

### CODE REVIEW

File: controller/controller.go [hunk 1]
```diff
@@ -698,6 +704,96 @@ func (s *Service) heartbeatController(ctx context.Context, addr *url.URL) {
 	}
 }
 
+func (s *Service) watchModuleChanges(ctx context.Context, sendChange func(response *ftlv1.PullSchemaResponse) error) error {
```
Comment (wesbillman) [inline controller/controller.go:707]: @alecthomas this seems like a lot of code to achieve what we want here :)

Any thoughts on how to refactor this a bit so it's not so much code? or does this look ok?

File: protos/xyz/block/ftl/v1/ftl.proto [hunk 2]
```diff
@@ -51,12 +51,20 @@ service VerbService {
   rpc Call(CallRequest) returns (CallResponse);
 }
 
+enum DeploymentChangeType {
+  DEPLOYMENT_ADDED = 0;
+  DEPLOYMENT_REMOVED = 1;
+  DEPLOYMENT_CHANGED = 2;
+}
```
Comment (wesbillman) [inline protos/xyz/block/ftl/v1/ftl.proto:58]: @alecthomas add this to support notifying when items are removed from the schema. I'm not sure if this was working, but curious to get your thoughts on this approach.

File: controller/controller.go (hunk 1 above)
Comment (alecthomas) [reply controller/controller.go:707]: You shouldn't need two loops. Get rid of the first one and just have the second one, that will diff against an empty map.

File: controller/controller.go [hunk 3]
```diff
@@ -698,6 +704,96 @@ func (s *Service) heartbeatController(ctx context.Context, addr *url.URL) {
...
+					ChangeType:    ftlv1.DeploymentChangeType_DEPLOYMENT_ADDED,
+				})
+			}
+			moduleSchemas[deployment.Schema.Name] = *deployment.Schema
+		}
+
+		var keysToRemove []string
+		for name, moduleSchema := range moduleSchemas {
+			if dbDeployment, ok := dbDeployments[name]; !ok {
+				keysToRemove = append(keysToRemove, name)
```
Comment (alecthomas) [inline controller/controller.go:762]: You can delete keys while iterating over the map - Go will handle this correctly.

File: controller/controller.go [hunk 4]
```diff
@@ -698,6 +704,96 @@ func (s *Service) heartbeatController(ctx context.Context, addr *url.URL) {
...
+
+		for _, deployment := range deployments {
+			dbDeployments[deployment.Schema.Name] = deployment
+			if current, ok := moduleSchemas[deployment.Schema.Name]; ok {
+				if !proto.Equal(current.ToProto(), deployment.Schema.ToProto()) {
+					//nolint:forcetypeassert
+					changesToSend = append(changesToSend, &ftlv1.PullSchemaResponse{
+						DeploymentKey: deployment.Key.String(),
+						Schema:        deployment.Schema.ToProto().(*pschema.Module),
+						ChangeType:    ftlv1.DeploymentChangeType_DEPLOYMENT_CHANGED,
```
Comment (alecthomas) [inline controller/controller.go:745]: Nice.

File: controller/controller.go [hunk 5]
```diff
@@ -698,6 +704,96 @@ func (s *Service) heartbeatController(ctx context.Context, addr *url.URL) {
...
+
+		for _, key := range keysToRemove {
+			delete(moduleSchemas, key)
+		}
+
+		for i, change := range changesToSend {
+			err := sendChange(&ftlv1.PullSchemaResponse{
+				DeploymentKey: change.DeploymentKey,
+				Schema:        change.Schema,
+				More:          i < len(changesToSend)-1,
```
Comment (alecthomas) [inline controller/controller.go:780]: You could potentially get rid of `changesToSend` entirely, and just `sendChange()` directly. This wouldn't batch new changes (with `More`), but in practice I don't think that's going to be that common.

File: controller/controller.go (hunk 1 above)
Comment (wesbillman) [reply controller/controller.go:707]: Good point! I had a different implementation before which required it, but the refactor helped with that :) Will update.

**Pull request merged**